src/Symfony/Component/Lock/Store/MongoDbStore.php

```diff
@@ -0,0 +1,426 @@
+                'Failed to put off the expiration of the "%s" lock within the specified time.',
+                $key
+            ));
+        }
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function delete(Key $key)
+    {
+        $filter = [
+            '_id' => (string) $key,
+            'token' => $this->getUniqueToken($key),
+        ];
+
+        $options = [];
+
+        $this->getCollection()->deleteOne($filter, $options);
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function exists(Key $key): bool
+    {
+        $filter = [
+            '_id' => (string) $key,
+            'token' => $this->getUniqueToken($key),
+            'expires_at' => [
+                '$gt' => $this->createDateTime(),
+            ],
+        ];
+
+        $doc = $this->getCollection()->findOne($filter);
+
+        return null !== $doc;
+    }
+
+    /**
+     * Update or Insert a Key.
+     *
+     * @param Key   $key
+     * @param float $ttl Expiry in seconds from now
+     */
+    private function upsert(Key $key, float $ttl)
+    {
+        $now = microtime(true);
+        $token = $this->getUniqueToken($key);
+
+        $filter = [
+            '_id' => (string) $key,
+            '$or' => [
+                [
+                    'token' => $token,
+                ],
+                [
+                    'expires_at' => [
+                        '$lte' => $this->createDateTime($now),
+                    ],
+                ],
+            ],
+        ];
+
+        $update = [
+            '$set' => [
+                '_id' => (string) $key,
+                'token' => $token,
+                'expires_at' => $this->createDateTime($now + $ttl),
+            ],
+        ];
+
+        $options = [
+            'upsert' => true,
+        ];
+
+        $this->getCollection()->updateOne($filter, $update, $options);
+    }
+
+    /**
+     * Remove expired locks from the collection. This is a fallback for MongoDB
+     * servers prior to version 2.2. For versions 2.2+ self::createTtlIndex is
+     * called instead based on the gcProbablity options.
+     */
+    private function removeExpiredLocks()
+    {
+        $now = microtime(true);
+
+        $filter = [
+            'expires_at' => [
+                '$lte' => $this->createDateTime($now),
+            ],
+        ];
+
+        $this->getCollection()->deleteMany($filter);
+    }
+
+    /**
+     * @return bool
+     */
+    private function isDuplicateKeyException(WriteException $e): bool
+    {
+        $writeErrors = $e->getWriteResult()->getWriteErrors();
+        if (1 === \count($writeErrors)) {
+            $code = $writeErrors[0]->getCode();
+        } else {
+            $code = $e->getCode();
+        }
+        // Mongo error E11000 - DuplicateKey
+        return 11000 === $code;
+    }
+
+    /**
+     * @return string
+     */
+    private function getDatabaseVersion(): string
+    {
+        if (null === $this->databaseVersion) {
```

Comment:
if (null !== $this->databaseVersion) {
return $this->databaseVersion;
}

$command = ....

Reply:
updated